### PR TITLE
feat: enable root span route instrumentation without any express layer spans 

### PR DESCRIPTION
**Pull request by shyimo**
### Which problem is this PR solving?
---
In addition to https://github.com/open-telemetry/opentelemetry-js-contrib/pull/176 PR, We should able to instrument the root span name when ignoring all express layers (router, middleware and request_handler).
This significantly reduces the amount of spans for a large and complex express based micro-service with large amount of outgoing http requests for a single incoming http request and still able to set the route on the root span. 

### Short description of the changes
---
using express with only request_handler layer before the change:
Screen Shot 2020-12-04 at 11 25 16

using express without request_handler, middleware and route layer (root span name not modified) before the change:
Screen Shot 2020-12-04 at 11 28 59

using express without request_handler, middleware and route layer after the change:
Screen Shot 2020-12-04 at 11 30 19

### CODE REVIEW

File: plugins/node/opentelemetry-plugin-express/src/express.ts [hunk 1]
```diff
@@ -180,23 +180,33 @@ export class ExpressPlugin extends BasePlugin<typeof express> {
...
+        // once we reach the request handler
         if (
           metadata.attributes[AttributeNames.EXPRESS_TYPE] ===
           ExpressLayerType.REQUEST_HANDLER
         ) {
           const parent = plugin._tracer.getCurrentSpan();
-          if (parent) {
-            parent.updateName(`${req.method} ${route}`);
+          // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+          // @ts-ignore
```
Comment (dyladan) [inline plugins/node/opentelemetry-plugin-express/src/express.ts:192]: Why is this ts ignore here?

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts [hunk 2]
```diff
@@ -102,6 +102,8 @@ describe('Express Plugin', () => {
       await tracer.withSpan(rootSpan, async () => {
         await httpRequest.get(`http://localhost:${port}/toto/tata`);
         rootSpan.end();
+        // @ts-ignore
```
Comment (dyladan) [inline plugins/node/opentelemetry-plugin-express/test/express.test.ts:105]: Why ts ignore?

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts [hunk 3]
```diff
@@ -204,6 +206,65 @@ describe('Express Plugin', () => {
...
+      });
+      const server = http.createServer(app);
+      await new Promise(resolve => server.listen(0, resolve));
+
+      const port = (server.address() as AddressInfo).port;
+      assert.strictEqual(memoryExporter.getFinishedSpans().length, 0);
+      await tracer.withSpan(rootSpan, async () => {
+        await httpRequest.get(`http://localhost:${port}/toto/tata`);
+        rootSpan.end();
+        // @ts-ignore
```
Comment (dyladan) [inline plugins/node/opentelemetry-plugin-express/test/express.test.ts:244]: ts ignore?

**Review by dyladan — COMMENTED** [review]
This is a very different behavior but it doesn't look like you added any configuration options. Am I missing something?

**Comment by shyimo** [comment]
Hi @dyladan.
Thats true, i didn't add any new configuration to the plugin.
The change only verify that will modify the root span name with the current request path (for example: HTTP POST -> POST user/:userid) when not using the request_handler express layer.
currently, the root span name is modified only when request_handler layer is on and this adds many tiny spans that sometimes is not needed

File: plugins/node/opentelemetry-plugin-express/src/express.ts (hunk 1 above)
Comment (shyimo) [reply plugins/node/opentelemetry-plugin-express/src/express.ts:192]: 'name' Property does not exist on the original type 'Span'. the name is added later on (in http instrumentation i think).
If i will not use ts-ignore annotation i will not pass the linter verification

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts (hunk 2 above)
Comment (shyimo) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts:105]: same as above

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts (hunk 3 above)
Comment (shyimo) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts:244]: same as above

File: plugins/node/opentelemetry-plugin-express/src/express.ts (hunk 1 above)
Comment (obecny) [reply plugins/node/opentelemetry-plugin-express/src/express.ts:192]: You create a new interface that extends api.Span and adds just optional name, we can't have `@ts-ignore` in here.

File: plugins/node/opentelemetry-plugin-express/src/express.ts [hunk 4]
```diff
@@ -180,23 +180,33 @@ export class ExpressPlugin extends BasePlugin<typeof express> {
...
+          const parentSpanName = parent && parent.name;
+          if (parent && parentSpanName) {
```
[the remark this answers is not in the document]
Comment (shyimo) [reply plugins/node/opentelemetry-plugin-express/src/express.ts]: @obecny looks good. fixed

File: plugins/node/opentelemetry-plugin-express/src/express.ts (hunk 1 above)
Comment (shyimo) [reply plugins/node/opentelemetry-plugin-express/src/express.ts]: @obecny no problem. done

**Review by dyladan — APPROVED** [review]

File: plugins/node/opentelemetry-plugin-express/test/express.test.ts [hunk 5]
```diff
@@ -204,6 +206,63 @@ describe('Express Plugin', () => {
       });
       server.close();
     });
+    it('should ignore all ExpressLayerType based on config. root span name should be modified when route exists', async () => {
```
[the remark this answers is not in the document]
Comment (shyimo) [reply plugins/node/opentelemetry-plugin-express/test/express.test.ts:209]: @obecny sure. on it

**Comment by shyimo** [comment]
@obecny Hi. i fix your comments on the unit test structure

**Review by obecny — APPROVED** [review]
lgtm, thx for those changes

**Review by dyladan — APPROVED** [review]

**Pull request merged**